Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -0,0 +1,306 @@
+   */
+  typedef FixedArray<value_type, kFixedSize> array_type;
+  typedef typename array_type::const_iterator const_iterator;
+
+  typedef std::shared_ptr<const array_type> array_pointer;
+
+  /**
+   * Creates an empty ArraySortedMap.
+   */
+  explicit ArraySortedMap(const C& comparator = C())
+      : array_(EmptyArray()), key_comparator_(comparator) {
+  }
+
+  /**
+   * Creates an ArraySortedMap containing the given entries.
+   */
+  explicit ArraySortedMap(std::initializer_list<value_type> entries,
+                          const C& comparator = C())
+      : array_(), key_comparator_(comparator) {
+    assert(static_cast<size_type>(entries.size()) <= kFixedSize);
+
+    auto array = std::make_shared<array_type>();
+    array->append(entries.begin(), entries.end());
+    array_ = array;
+  }
+
+  /**
+   * Creates a new map identical to this one, but with a key-value pair added or
+   * updated.
+   *
+   * @param key The key to insert/update.
+   * @param value The value to associate with the key.
+   * @return A new dictionary with the added/updated value.
+   */
+  this_type insert(const K& key, const V& value) const {
+    value_type pair(key, value);
+
+    const_iterator current_end = end();
+    const_iterator pos = LowerBound(pair.first);
+    bool replacing_entry = false;
+
+    if (pos != current_end) {
+      // LowerBound found an entry where pos->first >= pair.first. Reversing the
+      // argument order here tests pair.first < pos->first.
+      bool keys_equal = !key_comparator_(pair.first, *pos);
+      if (keys_equal) {
+        const V& pos_value = pos->second;
+        const V& pair_value = pair.second;
+        if (pos_value == pair_value) {
+          return *this;
+        } else {
+          replacing_entry = true;
+        }
+      }
+    }
+
+    // Copy the segment before the found position. If not found, this is
+    // everything.
+    auto copy = std::make_shared<array_type>();
+    copy->append(begin(), pos);
+
+    // Copy the value to be inserted.
+    copy->append(std::move(pair));
+
+    if (replacing_entry) {
+      // Skip the thing at pos because it compares the same as the pair above.
+      ++pos;
+    } else {
+      // If inserting at the end or the key at pos is not equal to what we're
+      // inserting, then increase the size.
+      assert(size() < kFixedSize);
+    }
+
+    // Copy everything after pos (if anything).
+    copy->append(pos, current_end);
+    return wrap(copy);
+  }
+
+  /**
+   * Creates a new map identical to this one, but with a key removed from it.
+   *
+   * @param key The key to remove.
+   * @return A new dictionary without that value.
+   */
+  this_type erase(const K& key) const {
+    const_iterator current_end = end();
+    const_iterator pos = find(key);
+    if (pos == current_end) {
+      return *this;
+    } else {
+      auto copy = std::make_shared<array_type>();
+      copy->size_ = size() - 1;
```

Comment:
Optional: consider:

```
auto copy = std::make_shared<array_type>(begin(), pos);
copy->append(pos + 1, current_end);
```

The only downside I see is that size is calculated twice. However, it's shorter and avoids accessing FixedArray internals (I think friendship would become unnecessary).